ext/drsyscall/drsyscall_record_lib.h

```diff
@@ -38,51 +38,106 @@
 #include "drsyscall.h"
 #include "drsyscall_record.h"
 
+/**
+ * A user provoided function to read syscall records. Returns the number of bytes read.
+ * Returns 0 if there are no more records.
+ */
+typedef size_t (*drsyscall_record_read_t)(DR_PARAM_IN char *buffer,
+                                          DR_PARAM_IN size_t size);
+
+/**
+ * A user provoided function to write syscall records. Returns the number of bytes
+ * written.
+ */
+typedef size_t (*drsyscall_record_write_t)(DR_PARAM_IN char *buffer,
+                                           DR_PARAM_IN size_t size);
+
+/**
+ * Callback function to invoke for each syscall record. Returns true to continue, false to
+ * stop.
+ */
+typedef bool (*drsyscall_iter_record_cb_t)(DR_PARAM_IN syscall_record_t *record);
+
+/**
+ * Callback function to invoke for each memory region. The callback function
+ * might be invoked multiple times for a memory region. If \p last_buffer is
+ * false, it indicates more data is available, true otherwise.
+ */
+typedef bool (*drsyscall_iter_memory_cb_t)(DR_PARAM_IN char *buffer,
+                                           DR_PARAM_IN size_t size,
+                                           DR_PARAM_IN bool last_buffer);
+
 DR_EXPORT
 /**
- * Reads system call records from \p record_file and outputs the content to \p
- * output_stream. Any errors encountered during this process are reported to the
- * \p error_stream.
- * @return 0 on success, or -1 if an error occurs.
+ * Dynamically iterates over all syscall records.
+ *
+ * @param[in] read_func  A user provided function to read syscall records.
+ * @param[in] record_cb  The callback to invoke for each syscall record.
+ * @param[in] memory_cb  The callback to invoke for memory region content. The
+ *                       callback may be invoke multiple times for a memory
+ *                       region. last_buffer is set to true for the last call of
+ *                       a memory region.
+ *
+ * \return true on success, or false if an error occurs.
```

Comment:
But the callbacks just say "false to stop": they don't imply it's an error; just a short-circuit.  But this iterate function returns false meaning an error if a callback returns false.  Please change the code or clarify the docs.